Cookbook/Proposals/AdditionalRulesForDanger.md

```diff
@@ -0,0 +1,133 @@
+
+Given that NVL is scheduled for release within two weeks, using NVL components is pivotal to Babylon's overall look and feel.
+With the exception of the empty space component (even though it has its own NVL equivalent), every other pre-NVL component used inside Renderer should trigger a warning.
+
+The following regex currently yields 1333 results in 216 files: `Component\.(?!EmptySpace)\b`.
+The sheer number of warnings immediately excludes SwiftLint; Danger would be our only option.
+
+Nevertheless, this would require that everyone relied exclusively on NVL components and I am not sure this is the case right now.
+
+
+### Mutating Current in `setUp()` but not restoring it in `tearDown()`
+
+The rule is simple: every object (susceptible to global mutation) that is mutated in `setUp()` must be restored in `tearDown()`.
+Mutations in `Current` in particular are the prime suspects of a number of flaky tests that have been recently affecting our test suite.
+
+In order to detect this problem we have to use Danger since SwiftLint only accepts regular expressions.
+We would have to whitelist every test file (`.*Tests.*`) and use Danger to manually parse the test line-by-line until the `setUp()` function was found. 
+
+If found, we need to make a counter that increases by one each time a curly bracket and does the reverse when a closing curly bracket appears. Once the counter reaches zero again, then we've successfully delimited the contents of the `setUp()` function and we'd use the following regex to detect changes in Current: `Current\..*\ = `.
+
+The algorithm is then be executed once more but for the `tearDown()` method instead.
+If the number of changes inside each function matches, then no issue would be found. If not, then Current appeared to have been mutated but not restored; we'd issue a warning for the lines in which the algorithm detected a change and alert the developers in GitHub.
+
+Nevertheless, this rule is rather naïve because we have no actual way of detecting if Current was reset or not; that would require compiling and running the actual Swift code. Therefore, this approach is not a silver bullet or anything remotely similar but it would still be useful for one use case in particular: should the developer forget to restore Current, this rule will alert us of this situation.
+
+### Translation errors
+
+Translations are particularly hard to catch during PR code reviews, especially if a lot of new phrases are added at once.
+While, in theory, we have a fastlane set up to deal with these translations there is always the risk that developers mistype the localization key when setting up a new static var.
+
+As such, I'd like to suggest another rule for Danger that would:
+    1. detect new static String variables in every Localizable.swift file.
+    2. if we detected a new static variable we would then scan the localization string file (`Babylon/Supporting Files/Localization/en.lproj/Localizable.strings`) and look for that key
```

Comment:
1. I think we need to do that in both directions. A key in `Localization*.swift` but not present in `Localizable.strings` is the biggest problem, but a key in the `.strings` file without a constant in the `.swift` file also means that we pulled a strings from Lokalise that we in fact don't use in this PR and is probably part of a different, unrelated feature, so we should warn about this too
2. What about other `.strings` files, especially `TargetSpecificLocalizable.strings`?

Also, I think we need to address the problem about handling the L10n across our modules and in tests. e.g. if that's what we've decided today (I'm not even sure what the actual concensus is today tbh), ensure that any reference to a `Localizable.strings` in a test target of a project file is just a reference to the `Localizable.strings` file of Babylon, so containing the same keys since it's the same file, and is not a separate `.strings` file where keys have been copy/pasted instead.

I think all that might even require a TechnicalDocument on its own jsut to describe how to do L10n on Babylon (using the dedicated lane to pull translations, clean up translations unrelated to the PR, in which files to add the translations, in which file to add the constants, depending on the feature they're related to (`LocalizationUI.swift` vs `LocalizationOctopus.swift`, etc), how to reference localized strings from tests by switching the bundle...

And once all those rules are clearly written black on white, the Danger rule will become much clearer too. e.g. I can already see a new rule we should have, similar to the one about `Current`, to ensure the bundle is reset in `tearDown` if it was changed in `setUp`...

Reply:
I'm all for using a code generation tool for that.

But we'll need a lot of cleanup of our existing codebase first I think, since as of now, the name of the constants we manually create and the corresponding Localization.strings keys they point to don't match names. So when we'll replace that with codegen, the name of our constants will completely change to match the keys used in the `.strings` and we'll have a lot of updating to do on all the code using them. Which means it's not a trivial task given our legacy and number of existing keys (we might consider a progressive transition though)

Since idk when we'll get there (transition to codegen), that might be an intermediate rule worth having in the meantime

Reply:
we'll have to do that anyway. if we need an intermediate solution that will help with it so that we can do that gradually rather than all in one go I think it's good. But again even if we start to code generate these strings today I don't think anything enforces us to use generated code all in, we can keep existing structures and just mark it as deprecated. Lots of options here.